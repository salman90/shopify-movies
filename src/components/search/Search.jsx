import React, { PureComponent } from 'react';
import styles from './Search.module.css';

class Search extends PureComponent {
    state = {
        text: '',
    }


    handleTextChange = (event) => {
        console.log(event.target.value)
        this.setState({
            text: event.target.value
        });

    }

    handleSubmitSearch = (event) => {
        event.preventDefault();
        this.props.handleSubmitSearch(this.state.text)
    }


    render(){
        return (
            <div
             className={styles.searchContainer}
            >
                <form className={styles.formStyle} onSubmit={this.handleSubmitSearch}>
                    <input className={styles.textInput} type="text" name="search" value={this.state.text} onChange={this.handleTextChange}/>
                    <input type="submit" name='search' value="submit" className={styles.buttonStyle} />
                </form>
            </div>
        )
    }
}


export default Search