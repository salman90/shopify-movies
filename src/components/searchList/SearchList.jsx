import React, {PureComponent} from 'react';
import SearchItem from './searchItem';
import Spinner from '../../components/spinner';
import styles from './SearchList.module.css';


/**
 * @class SearchList
 * @extends React.PureComponent
 * @description this is a class contains all the seachItems
 */
class SearchList extends PureComponent {

    handelNomination = (nomnationItem) => {
        this.props.handelNomination(nomnationItem);
    }

    /**
     * @description traverse each searchItem
     * @param {Array} searchItems 
     * @returns {React.PureComponent} SearchItem
     */
    renderSearchItem = (searchItems) => (
        searchItems.map((searchItemDetails) => {
            return (
                    <SearchItem
                        key={searchItemDetails.imdbID}
                        searchItemDetails={searchItemDetails}
                        handelNomination={this.handelNomination}
                    />
            )
        })
    )   

    /**
     * 
     */
    renderLoading = () => {
        if(!this.props.loading)
        {
            return null;
        }
        else
        {
            return (<Spinner />)
        }
    }

    render() {
        return( 
            <div className={styles.searchListContainer}>
                <p className={styles.searchTitle} style={{textAlign: 'center'}}>Search Results</p>
                {this.props.searchResult.length > 0 ? 
                  this.renderSearchItem(this.props.searchResult)
                  :
                 this.renderLoading()
                 }
            </div>
        )
    }
}

export default SearchList;