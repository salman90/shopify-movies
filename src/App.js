import React, { PureComponent } from 'react';
import axios from 'axios';
import appStyles from './App.module.css';
import Header from './components/header';
import Search from './components/search';
import SearchList from './components/searchList';
import NominationList from './components/nomination';
import { ToastContainer, toast } from 'react-toastify';
import { URL, API_KEY, PARAM_TYPE } from './constants';
import 'react-toastify/dist/ReactToastify.css';
import { removeFromArray } from './util'

/**
 * @class App
 * @extends React.PureComponent
 * @description this a component of the all the components and it handle all the application state
 */
class App extends PureComponent {
/**
 * @constructor
 * @param {object} props
 */
  constructor(props) {
    super(props)
    this.state = {
      searchResult: [],
      nominations: [],
    }
  }


  componentDidMount() {
    const nominations = JSON.parse(localStorage.getItem('nominations'));
    if (nominations !== null && nominations.length > 0)
    {
      this.setState({
        nominations: nominations
      })
    }
  }

  componentDidUpdate(prevProps, prevState, snapshot){
    if (prevState.nominations.length !== this.state.nominations.length)
    {
      localStorage.setItem('nominations', JSON.stringify(this.state.nominations));
    }
  }


  
  /**
   * @description fech information from Omdb API and sets the searchResult to data coming from the API
   * @param {String} text - seachInput
   */
  handleSubmitSearch = async (text) => {
   if(text.length > 0){
     this.fetchMovies(text)
   }
   else
   {
     toast.warn("Serach is empty")
   }
  }

  async fetchMovies(text){
    axios.get(`${URL}apikey=${API_KEY}`, {
      params: {
        s: text,
        type: PARAM_TYPE
      }
    })
    .then((res) => {
      if (res.data.Response === "True")
      {
        this.setState({
          searchResult: res.data.Search
        })
      }
      else
      {
        throw new Error(res.data.Error);
      }
    })
    .catch((err) => {
      toast.error(err.toString())
    })
  }

  /**
   * @description adds a nominationItem Object to nominations array
   * @param {Object} nominationItem
   */
  addNominations = (nominationItem) => {
    if (this.state.nominations.some(nomination => nomination.imdbID === nominationItem.imdbID))
    {
      toast.warn(nominationItem.Title + " already exists in nominations");
    }
    else if (this.state.nominations.length >= 5)
    {
      toast.warn("Can't add more than five nominations")
    }
    else
    {
      const newArray = removeFromArray(nominationItem, this.state.searchResult);
      if (newArray !== null)
      {
        this.setState({
          nominations: [...this.state.nominations, nominationItem],
          searchResult: newArray
        });
        toast.success(nominationItem.Title + " added to nominations list successfully");
      }
    }
  }


  /**
   * @description removes a nominationItem from nomination array
   * @param {Object} nominationItem 
   */
  removeNomination = (nominationItem) => {
    const newArray = removeFromArray(nominationItem, this.state.nominations);
    
    if (newArray !== null)
    {
      this.setState({ nominations: newArray });
      toast.success("Removed " + nominationItem.Title + " from nomination list")
    }
    else
    {
      toast.error("can't remove " + nominationItem.Title)
    }

    
  }


  render(){
    return (
      <div className={appStyles.App}>
        <Header />
        <Search
          handleSubmitSearch={this.handleSubmitSearch}
        />
        <div
          className={appStyles.ListsContainer}
        >
          <SearchList
            searchResult={this.state.searchResult}
            handelNomination={this.addNominations}
          />
          <NominationList
            nominations={this.state.nominations} 
            removeNomination={this.removeNomination}
          />
        </div>
        <ToastContainer 
          draggable={false}
          position={'bottom-right'}
          autoClose={3000}
        />
      </div>
    );
  }
}

export default App;
