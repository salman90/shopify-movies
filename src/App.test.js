import React from 'react';
// import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { mount, shallow } from 'enzyme';

import App from './App';
import Search from './components/search';
import SearchList from './components/searchList'

// test('renders learn react link', () => {
//   const { getByText } = render(<App />);
//   const linkElement = getByText(/learn react/i);
//   expect(linkElement).toBeInTheDocument();
// });


// test


describe('App', () => {
  
  test('snapshot renders', () => {

    const component = renderer.create(<App />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('render the orginal loading state', () => {
    const wrapper = shallow(<App />);
    expect(wrapper.state('loading')).toBe(false);
  })


  it('renders child Search component', () => {
    // const wrapper =  mount(<App />);
    // expect(wrapper.find(Search).length.toEqual(1));
    const wrapper = mount(<App />);
    expect(wrapper.find(Search).length).toEqual(1);
  });
  
});
