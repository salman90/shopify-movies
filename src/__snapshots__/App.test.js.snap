// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App snapshot renders 1`] = `
<div
  className="App"
>
  <header
    className="AppHeader"
  >
    <div
      className="LogoContainer"
    >
      <img
        alt="shopify logo"
        src="shopifyLogo.png"
        style={
          Object {
            "height": 56,
            "width": 56,
          }
        }
      />
      <p
        className="logoText"
      >
        shopify
        <span
          className="spantext"
        >
          Movies
        </span>
      </p>
    </div>
  </header>
  <div
    className="searchContainer"
  >
    <form
      className="formStyle"
      onSubmit={[Function]}
    >
      <div
        className="MuiFormControl-root MuiTextField-root textInput"
      >
        <div
          className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl MuiInputBase-adornedStart MuiOutlinedInput-adornedStart"
          onClick={[Function]}
        >
          <div
            className="MuiInputAdornment-root MuiInputAdornment-positionStart"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
              />
            </svg>
          </div>
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedStart MuiOutlinedInput-inputAdornedStart"
            disabled={false}
            id="input-with-icon-textfield"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="text"
            value=""
          />
          <fieldset
            aria-hidden={true}
            className="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
            style={
              Object {
                "paddingLeft": 8,
              }
            }
          >
            <legend
              className="PrivateNotchedOutline-legend-2"
              style={
                Object {
                  "width": 0.01,
                }
              }
            >
              <span
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "&#8203;",
                  }
                }
              />
            </legend>
          </fieldset>
        </div>
      </div>
      <button
        className="buttonStyle"
        name="search"
        type="submit"
        value="submit"
      >
        Search
      </button>
    </form>
  </div>
  <div
    className="ListsContainer"
  >
    <div
      className="searchListContainer"
    >
      <p
        className="searchTitle"
      >
        Search Results
      </p>
    </div>
    <div
      className="nominationContainer"
    >
      <div
        className="nominationListTitleContainer"
      >
        <p
          className="nominationTitle"
        >
          Nominations
        </p>
      </div>
      <div
        className="nominationItemContainer"
      />
    </div>
  </div>
  <div
    className="Toastify"
  />
</div>
`;
